k8s/client.go

```diff
@@ -62,46 +89,198 @@ func New(apiServerURL string, caCert []byte, token string) (Client, error) {
+		}
+
+		retryRequest := func() (*http.Response, error) {
+			return retryRequest(w, ingressRequest)
+		}
+
+		for watch(w, retryRequest) {
+		}
+
+		log.Debug("Watch has stopped")
+	}()
+
+	return nil
+}
+
+type genericList struct {
+	ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+}
+
+func (c *client) getIngressResourceVersion() (string, error) {
+	var ingresses genericList
+	err := c.unmarshallRequest(ingressPath, &ingresses)
 	if err != nil {
-		return nil, err
+		return "", err
+	}
+	return ingresses.ResourceVersion, nil
+}
+
+// watch returns true if it should be retried, false if the watcher has terminated.
+func watch(w Watcher, request func() (*http.Response, error)) bool {
+	resp, err := request()
+	if err != nil {
+		log.Infof("Watcher could not make request, shutting down: %v", err)
+		return false
 	}
 	defer resp.Body.Close()
+	log.Infof("Watching %v", resp.Request.URL)
 
-	if resp.StatusCode < 200 || 300 <= resp.StatusCode {
-		return nil, fmt.Errorf("GET %s returned %v", endpoint, resp)
+	// send an update for a successful watch start
+	w.Updates() <- struct{}{}
+
+	updateCh := make(chan interface{})
+
+	go func() {
+		defer close(updateCh)
+
+		scanner := bufio.NewScanner(resp.Body)
+
+		for scanner.Scan() {
+			event := scanner.Text()
+			log.Infof("Received event %s", event)
+			updateCh <- struct{}{}
+		}
+
+		if err := scanner.Err(); err != nil {
+			log.Debugf("Error while watching events, closing update channel: %v", err)
+		}
+	}()
+
+	for {
+		var reconnect bool
+
+		select {
+		case <-w.Done():
+			log.Debug("Watcher is done, stopping watch")
+			return false
+		case update := <-updateCh:
+			if update == nil {
+				log.Info("Long poll terminated, will reconnect")
+				reconnect = true
+			} else {
+				log.Debug("Noticed update, sending notification to watcher")
+				w.Updates() <- update
+			}
+		}
+
+		if reconnect {
+			break
+		}
+	}
+
+	return true
+}
+
+func retryRequest(w Watcher, request func() (*http.Response, error)) (*http.Response, error) {
+	respCh := make(chan *http.Response)
+	delay := initialRetryDelay
+	go func() {
+		defer close(respCh)
+		var jitter = rand.New(rand.NewSource(time.Now().UnixNano()))
+		t := time.NewTimer(0)
+
+		for {
+			select {
+			case <-w.Done():
+				log.Debug("Watcher is closed, stopping retry")
+				return
+			case <-t.C:
+				resp, err := request()
+
+				if err != nil {
+					delay += time.Millisecond * time.Duration(jitter.Intn(50))
+					log.Warnf("Failed to request watch, will retry in %v: %v", delay, err)
+					t.Reset(delay)
+					if delay < maxRetryDelay {
+						delay = delay * 2
+					}
+					break
+				}
+
+				log.Debugf("Succeeded watching %v", resp.Request.URL)
+				respCh <- resp
+				return
+			}
+		}
+	}()
+
+	resp := <-respCh
+	if resp == nil {
+		return nil, fmt.Errorf("request shutdown")
 	}
+	return resp, nil
+}
+
+func (c *client) unmarshallRequest(path string, val interface{}) error {
```

Comment:
This also makes the request where as from the name I expected it to just take a Http response and unmarshall it. Would it make sense to inline request into this method and rename this one to request?